src/hotspot/share/gc/serial/serialHeap.cpp

```diff
@@ -756,45 +756,30 @@ void SerialHeap::process_roots(ScanningOption so,
   }
 }
 
-bool SerialHeap::no_allocs_since_save_marks() {
-  return _young_gen->no_allocs_since_save_marks() &&
-         _old_gen->no_allocs_since_save_marks();
-}
-
 template <typename OopClosureType>
-static void oop_iterate_from(OopClosureType* blk, ContiguousSpace* space, HeapWord* from) {
-  assert(from != nullptr, "precondition");
+static void oop_iterate_from(OopClosureType* blk, ContiguousSpace* space, HeapWord** from) {
+  assert(*from != nullptr, "precondition");
   HeapWord* t;
-  HeapWord* p = from;
+  HeapWord* p = *from;
 
   const intx interval = PrefetchScanIntervalInBytes;
   do {
     t = space->top();
     while (p < t) {
       Prefetch::write(p, interval);
-      oop m = cast_to_oop(p);
-      p += m->oop_iterate_size(blk);
+      p += cast_to_oop(p)->oop_iterate_size(blk);
     }
   } while (t < space->top());
+
+  *from = space->top();
 }
 
 void SerialHeap::scan_evacuated_objs(YoungGenScanClosure* young_cl,
                                      OldGenScanClosure* old_cl) {
   do {
-    {
-      HeapWord* saved_top = young_gen()->saved_mark_word();
-      oop_iterate_from(young_cl, young_gen()->to(), saved_top);
-      young_gen()->set_saved_mark_word();
-    }
-    {
-      HeapWord* saved_top = old_gen()->saved_mark_word();
-      oop_iterate_from(old_cl, old_gen()->space(), saved_top);
-      old_gen()->set_saved_mark_word();
-    }
-    if (young_gen()->no_allocs_since_save_marks()) {
-      break;
-    }
-  } while (true);
+    oop_iterate_from(young_cl, young_gen()->to(), &_young_gen_saved_top);
+    oop_iterate_from(old_cl, old_gen()->space(), &_old_gen_saved_top);
+  } while (_young_gen_saved_top != young_gen()->to()->top());
```

Comment:
Shouldn't this condition also consider changes to `_old_gen_saved_top`?

If an iteration only moves to the old gen (e.g. to-space is full, or just all objects are of the appropriate age), these newly copied objects ought to also be scanned again. Or is there a reason that this is not necessary?

Also the statement `young_gen()->to()` could be factored out into a local to remove one indirection level in the loop.